File: R/oro.R
# Cargamos las librerías necesarias
library(rdbnomics)
library(tidyverse)
library(lubridate)
library(scales)

# 1. Obtener datos de precio del oro desde DBnomics
# Provider: LBMA, Dataset: gold_M
df_oro_raw <- rdb(
  provider_code = 'LBMA',
  dataset_code = 'gold_M',
  mask = 'M.EUR.PM.AVG'
)

# 2. Procesar datos del oro
df_oro <- df_oro_raw %>%
  filter(period >= as.Date("2000-01-01")) %>%
  select(period, value) %>%
  drop_na()

# Gráfico: Precio Nominal
ggplot(df_oro, aes(x = period, y = value)) +
  geom_line(color = "blue") +
  geom_point(size = 1) +
  labs(title = "Precio mensual del oro (EUR) desde 2000",
       x = "Fecha", y = "Precio del oro (EUR)") +
  theme_minimal()

# 3. Descargar y procesar datos del IPC (INE España)
url_ipc <- "https://www.ine.es/jaxiT3/files/t/csv_bdsc/50911.csv"
df_ipc_raw <- read_delim(url_ipc, delim = ";", locale = locale(decimal_mark = ","))

df_ipc <- df_ipc_raw %>%
  filter(`Tipo de dato` == "Variación mensual") %>%
  mutate(
    # Reemplazamos 'M' por '-' para convertir a fecha
    period_str = str_replace(Periodo, "M", "-"),
    period = ym(period_str)
  ) %>%
  select(period, IPC = Total)

# 4. Merge y cálculos de precio real
df_oro_ipc <- inner_join(df_oro, df_ipc, by = "period") %>%
  arrange(period)

# Calcular índice IPC y valor real
df_oro_ipc <- df_oro_ipc %>%
  mutate(
    # El primer valor del índice se basa en la primera variación
    IPC_indice = 100 * cumprod(1 + IPC / 100),
    value_real = value * (first(IPC_indice) / IPC_indice)
  )

# Gráfico: Nominal vs Real
ggplot(df_oro_ipc) +
  geom_line(aes(x = period, y = value, color = "Nominal")) +
  geom_line(aes(x = period, y = value_real, color = "Real (Ajustado IPC)")) +
  labs(title = "Precio mensual del oro: nominal vs real",
       x = "Fecha", y = "Precio (EUR)", color = "Tipo") +
  theme_minimal()

# 5. Incrementos y cambios acumulados
df_oro_ipc <- df_oro_ipc %>%
  mutate(
    real_monthly_change = (value_real / lag(value_real) - 1) * 100,
    real_cumulative_change = (value_real / first(value_real) - 1) * 100,
    real_annualized_change = ((1 + real_monthly_change / 100)^12 - 1) * 100
  )

# Gráfico: Incremento acumulado
ggplot(df_oro_ipc, aes(x = period, y = real_cumulative_change)) +
  geom_line(color = "darkgreen") +
  labs(title = "Incremento acumulado real del precio del oro desde 2000",
       x = "Fecha", y = "Incremento acumulado (%)") +
  theme_minimal()

# 6. Cálculo de CAGR anual (por año)
cagr_per_year <- df_oro_ipc %>%
  group_by(year = year(period)) %>%
  summarise(
    valor_inicial = first(value_real),
    valor_final = last(value_real),
    n_months = n(),
    cagr_year = ((valor_final / valor_inicial)^(12 / n_months) - 1) * 100
  )

# Gráfico: CAGR por año
ggplot(cagr_per_year, aes(x = year, y = cagr_year)) +
  geom_col(fill = "steelblue") +
  labs(title = "CAGR anual del precio real del oro",
       x = "Año", y = "CAGR (%)") +
  theme_minimal()

# 7. CAGR total del periodo
valor_inicial_tot <- first(df_oro_ipc$value_real)
valor_final_tot <- last(df_oro_ipc$value_real)
n_years <- as.numeric(difftime(last(df_oro_ipc$period), first(df_oro_ipc$period), units = "days")) / 365.25

cagr_total <- ((valor_final_tot / valor_inicial_tot)^(1/n_years) - 1) * 100
cat(sprintf("Media anualizada del incremento real (CAGR): %.2f%%\n", cagr_total))

# Guardar CSV
write_csv(df_oro_ipc, "precio_oro_eur_r.csv")

File: Python/oro.py
#%%
"""
This script downloads, processes, and analyzes monthly gold prices in EUR and Spanish general CPI data, then computes the real gold price and its real monthly change since 2000.

Workflow:
1. Fetches monthly gold price data (in EUR, PM fixing, average price) from the LBMA dataset via DBnomics.
2. Filters the gold price data to include only records from January 2000 onwards and saves it as a CSV.
3. Plots the monthly gold price in EUR since 2000.
4. Downloads Spanish general CPI variation rates from INE, processes the date format, and saves as a CSV.
5. Merges gold price and CPI data on the period, computes the real gold price (adjusted for inflation), and calculates the real monthly percentage change in gold price.

Dependencies:
- dbnomics
- pandas
- matplotlib

Outputs:
- 'precio_oro_eur.csv': Gold price data since 2000.
- 'ipc_general.csv': CPI data.
- DataFrame with real gold price and its real monthly increment.
"""
from dbnomics import fetch_series
import pandas as pd
import matplotlib.pyplot as plt

#%%
# Definimos los parámetros de la serie
# Provider: LBMA, Dataset: gold_M
# Dimensions: se pasan como un diccionario de Python
df = fetch_series(
    provider_code='LBMA',
    dataset_code='gold_M',
    dimensions={
        'frequency': ['M'], 
        'unit': ['EUR'], 
        'time': ['PM'], 
        'price': ['AVG']
    }
)

#%%
# La librería ya devuelve un DataFrame con 'period' convertido a datetime
# Solo aplicamos el filtro de fecha
df_oro = df[df['period'] >= '2000-01-01'][['period', 'value']]

plt.figure(figsize=(10, 5))
plt.plot(df_oro['period'], df_oro['value'], marker='o', linestyle='-')
plt.xlabel('Fecha')
plt.ylabel('Precio del oro (EUR)')
plt.title('Precio mensual del oro (EUR) desde 2000')
plt.grid(True)
plt.tight_layout()
plt.show()



# %%
url = "https://www.ine.es/jaxiT3/files/t/csv_bdsc/50911.csv"  # tasa de variación IPC general [web:9]
df_ipc = pd.read_csv(url, sep=';', decimal=',')
df_ipc = df_ipc[df_ipc['Tipo de dato'] == 'Variación mensual'] # nos quedamos solo con la variación mensual

df_ipc['period'] = pd.to_datetime(df_ipc['Periodo'].str.replace('M', '-'), format='mixed')
df_ipc = df_ipc[['period', 'Total']].rename(columns={'Total': 'IPC'})
#%%
df_oro_ipc = pd.merge(df_oro, df_ipc, on='period', how='inner')
#%%
# Calcular el precio real del oro ajustado por IPC (base 100 en el primer mes)
ipc_base = df_oro_ipc['IPC'].iloc[0]
df_oro_ipc['IPC_indice'] = 100 * (1 + df_oro_ipc['IPC'] / 100).cumprod()
df_oro_ipc['value_real'] = df_oro_ipc['value'] * (df_oro_ipc['IPC_indice'].iloc[0] / df_oro_ipc['IPC_indice'])
plt.figure(figsize=(10, 5))
plt.plot(df_oro_ipc['period'], df_oro_ipc['value'], label='Precio nominal del oro (EUR)')
plt.plot(df_oro_ipc['period'], df_oro_ipc['value_real'], label='Precio real del oro (EUR, ajustado IPC)')
plt.xlabel('Fecha')
plt.ylabel('Precio del oro (EUR)')
plt.title('Precio mensual del oro: nominal vs real (ajustado por IPC)')
plt.legend()
plt.grid(True)
plt.tight_layout()
plt.show()

#%% 
# Calcular el incremento real en porcentaje mensual
df_oro_ipc['real_monthly_change'] = df_oro_ipc['value_real'].pct_change() * 100

# Calculalar el incremento real en procentaje acumulado desde el inicio
df_oro_ipc['real_cumulative_change'] = (df_oro_ipc['value_real'] / df_oro_ipc['value_real'].iloc[0] - 1) * 100

plt.figure(figsize=(10, 5))
plt.plot(df_oro_ipc['period'], df_oro_ipc['real_cumulative_change'], label='Incremento acumulado real (%)')
plt.xlabel('Fecha')
plt.ylabel('Incremento acumulado real (%)')
plt.title('Incremento acumulado real del precio del oro (ajustado por IPC) desde 2000')
plt.legend()
plt.grid(True)
plt.tight_layout()
plt.show()

#%%
# calcular la tasa anualizada del incremento real media
df_oro_ipc['real_annualized_change'] = ((1 + df_oro_ipc['real_monthly_change'] / 100) ** 12 - 1) *  100
plt.figure(figsize=(10, 5))
plt.plot(df_oro_ipc['period'], df_oro_ipc['real_annualized_change'], label='Tasa anualizada del incremento real (%)')
plt.xlabel('Fecha')
plt.ylabel('Tasa anualizada del incremento real (%)')
plt.title('Tasa anualizada del incremento real del precio del oro (ajustado por IPC)')
plt.legend()
plt.grid(True)
plt.tight_layout()
plt.show()

#%%
df_oro_ipc.to_csv('precio_oro_eur.csv', index=False)
# %%

#%% Calcular media anualizada del incremento real (CAGR) de cada año
df_oro_ipc['year'] = df_oro_ipc['period'].dt.year
cagr_per_year = {}
for year, group in df_oro_ipc.groupby('year'):
    valor_inicial = group['value_real'].iloc[0]
    valor_final = group['value_real'].iloc[-1]
    n_months = len(group)
    cagr_year = (valor_final / valor_inicial) ** (12 / n_months) - 1
    cagr_per_year[year] = cagr_year * 100


# plotear el CAGR por año
plt.figure(figsize=(10, 5))
plt.bar(cagr_per_year.keys(), cagr_per_year.values())
plt.xlabel('Año')
plt.ylabel('CAGR (%)')
plt.title('CAGR anual del precio real del oro (ajustado por IPC)')
plt.grid(True)
plt.tight_layout()
plt.show()

# Calcular la media anualizada (CAGR) del precio real del oro en todo el periodo
valor_inicial = df_oro_ipc['value_real'].iloc[0]
valor_final = df_oro_ipc['value_real'].iloc[-1]
n_years = (df_oro_ipc['period'].iloc[-1] - df_oro_ipc['period'].iloc[0]).days / 365.25

cagr = (valor_final / valor_inicial) ** (1 / n_years) - 1
print(f"Media anualizada del incremento real (CAGR): {cagr * 100:.2f}%")
# ...existing code...

# %%
